### PR TITLE
OpenSearch Index DateTime Pattern conversion

### LINKED ISSUE #854: LogstashConfigConverter: converting index with a date-time pattern
**Is your feature request related to a problem? Please describe.**
As the RFC #755, Data-Prepper not supports index names with a date-time pattern as a suffix. 
Logstash also supports a date-time pattern in index. However, Logstash doesn't support the date-time pattern conversion. 

**Describe the solution you'd like**
We should have Logstash Config Converter to be able to convert Logtash's version of index with data a time pattern to Data Prepper's version of index with a data-time pattern. 

One catch: 
Logstash's date time patten uses Joda time [formatter](https://www.joda.org/joda-time/key_format.html). 
Data-Prepper date time pattern uses Java Time [DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)

**Describe alternatives you've considered (Optional)**
NA
**Additional context**
NA

### PR DESCRIPTION
**Pull request by sshivanii**
### Description
* Adds a custom AttributeMapper to converts `year`, `year-of-era`, `week-year` patterns in index from [Joda](https://www.joda.org/joda-time/key_format.html) to [Java-time](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html):
* Removes `+` from DateTime patterns in Logstash config
* Adds test cases

 
### Issues Resolved
Resolves #854 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+        final String logstashIndexAttributeValue = logstashAttribute.getAttributeValue().getValue().toString();
+        final Pattern pattern = Pattern.compile(REGEX_EXTRACTING_DATETIME_PATTERN);
+        final Matcher dateTimePatternMatcher = pattern.matcher(logstashIndexAttributeValue);
+
+        final StringBuilder updatedDateTimePattern = new StringBuilder();
+        String updatedIndexAttributeValue = "";
+
+        if (dateTimePatternMatcher.find()) {
+            final String dateTimePattern = dateTimePatternMatcher.group(0);
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:52]: What if there are multiple index patterns in the config? I'm not sure if it happens much, but it might be worth considering. We could also fix this in a fast follow-on PR if that helps.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+class OpenSearchPluginAttributesMapperTest {
+
+    private OpenSearchPluginAttributesMapper createObjectUnderTest() {
+        return new OpenSearchPluginAttributesMapper();
+    }
+
+    @Test
+    void convert_logstash_index_date_time_pattern() {
+        final String dataPrepperIndexAttribute = "index";
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:32]: You can make this a constant in the class and use it elsewhere. It will make the code more readable since it will be clear that it is a constant.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        final Map<String, Object> pluginSettings =
+                createObjectUnderTest().mapAttributes(Collections.singletonList(logstashAttribute), logstashAttributesMappings);
+
+        assertThat(pluginSettings, notNullValue());
+        assertThat(pluginSettings.size(), equalTo(1));
+        assertThat(pluginSettings, hasKey(dataPrepperIndexAttribute));
+    }
+
+    @Test
+    void convert_logstashIndexPattern_joda_year_to_dataPrepperIndexPattern_java_year() {
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:54]: I think that this test and the next three are the same except for the input and output. You can make this a JUnit 5 parameterized test and this would be clearer to use.

Here is somewhat of a starting point for it.

```
    static class JodaToJava8IndicesArgumentsProvider implements ArgumentsProvider {
        @Override
        public Stream<? extends Arguments> provideArguments(final ExtensionContext extensionContext) {
            return Stream.of(
                    Arguments.arguments("logstash-%{+yyyy.MM.dd}", "logstash-%{uuuu.MM.dd}"),
                    Arguments.arguments("logstash-index-%{+YYYY.MM.dd}", "logstash-index-%{yyyy.MM.dd}"),
                    // repeat for each
            );
        }
    }
```


You can see this in use at:

https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/LogstashConfigConverterIT.java#L52-L54

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thanks, this loos good. I made a few suggestions that I think will help improve the code.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java (hunk 1 above)
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:52]: As far as I checked, there can be only `index` pattern per plugin at least so my assumption here is there's only one. Also, no nested patterns are allowed.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        final LogstashAttributesMappings logstashAttributesMappings = mock(LogstashAttributesMappings.class);
+        when(logstashAttributesMappings.getMappedAttributeNames()).thenReturn(Collections.singletonMap(LOGSTASH_OPENSEARCH_INDEX_ATTRIBUTE_NAME, dataPrepperIndexAttribute));
+
+
+        final Map<String, Object> pluginSettings =
+                createObjectUnderTest().mapAttributes(Collections.singletonList(logstashAttribute), logstashAttributesMappings);
+
+        assertThat(pluginSettings, notNullValue());
+        assertThat(pluginSettings.size(), equalTo(1));
+        assertThat(pluginSettings, hasKey(dataPrepperIndexAttribute));
```
[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:50]: Will add, I missed it in this test, thanks for catching this.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class OpenSearchPluginAttributesMapper extends AbstractLogstashPluginAttributesMapper {
+    protected static final String LOGSTASH_OPENSEARCH_INDEX_ATTRIBUTE_NAME = "index";
+    protected static final String REGEX_EXTRACTING_DATETIME_PATTERN = "%\\{(.*?)\\}";
```
[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:20]: That's a good point, I will add the `+` in the Regex.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    private String convertIndexDateTimePattern(LogstashAttribute logstashAttribute) {
+
+        final String logstashIndexAttributeValue = logstashAttribute.getAttributeValue().getValue().toString();
+        final Pattern pattern = Pattern.compile(REGEX_EXTRACTING_DATETIME_PATTERN);
+        final Matcher dateTimePatternMatcher = pattern.matcher(logstashIndexAttributeValue);
+
+        final StringBuilder updatedDateTimePattern = new StringBuilder();
+        String updatedIndexAttributeValue = "";
```
Comment (cmanning09) [inline data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:49]: As a best practice, variables should be minimally scoped to the block they are used and we should avoid re-assignment where possible. We can achieve this by leveraging an additional private method to handle the match and conversion part. 

```
return dateTimePatternMatcher.find() ? matchAndConvertDateTimePattern(dateTimePatternMatcher, logstashIndexAttributeValue) : logstashIndexAttributeValue;

```
New method:
```
private String matchAndConvertDateTimePattern(final Matcher dateTimePatternMatcher, final String logstashIndexAttributeValue) {

        final String dateTimePattern = dateTimePatternMatcher.group(0);
        final StringBuilder updatedDateTimePattern = new StringBuilder();
        for (final char character: dateTimePattern.toCharArray()) { 
           ... // conversion logic from below.
        }
        return logstashIndexAttributeValue.replace(dateTimePattern, updatedDateTimePattern);
    }
```

The current implementation also has a bug when a pattern is not provided because we are assigning the updatedIndexAttributeValue to `""` and returning `""` if a pattern does not exist.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    protected HashSet<String> getCustomMappedAttributeNames() {
+        return new HashSet<>(Collections.singletonList(LOGSTASH_OPENSEARCH_INDEX_ATTRIBUTE_NAME));
+    }
+
+    private String convertIndexDateTimePattern(LogstashAttribute logstashAttribute) {
+
+        final String logstashIndexAttributeValue = logstashAttribute.getAttributeValue().getValue().toString();
+        final Pattern pattern = Pattern.compile(REGEX_EXTRACTING_DATETIME_PATTERN);
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:45]: Yea, pattern means two things here: 1) regex and 2) a date-time pattern.

Maybe you can rename it to `dateTimeRegexPattern` to be clear in what it provides.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:52]: Thanks for clarifying on that. This is good then.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java (hunk 6 above)
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:49]: Thanks for catching this. It does substitute `""` for plain Strings index names. I will update this and add a test for this usecase.

Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:49]: @cmanning09 A few clarifications in your suggestion:
> As a best practice, variables should be minimally scoped to the block they are used and we should avoid re-assignment where possible. We can achieve this by leveraging an additional private method to handle the match and conversion part.
> 
* When you say additional private method, did you instead mean replacing the current `convertIndexDateTimePattern(LogstashAttribute logstashAttribute)` method for `matchAndConvertDateTimePattern(final Matcher dateTimePatternMatcher, final String logstashIndexAttributeValue` method?

> ```
> return dateTimePatternMatcher.find() ? matchAndConvertDateTimePattern(dateTimePatternMatcher, logstashIndexAttributeValue) : logstashIndexAttributeValue;
> ```
* I'm not sure I understand this return statement here, can you provide more context? 
> 
> New method:
> 
> ```
> private String matchAndConvertDateTimePattern(final Matcher dateTimePatternMatcher, final String logstashIndexAttributeValue) {
> 
>         final String dateTimePattern = dateTimePatternMatcher.group(0);
>         final StringBuilder updatedDateTimePattern = new StringBuilder();
>         for (final char character: dateTimePattern.toCharArray()) { 
>            ... // conversion logic from below.
>         }
>         return logstashIndexAttributeValue.replace(dateTimePattern, updatedDateTimePattern);
>     }
> ```
Based on the code snippet provided for the new method, are we okay with defining `Matcher` in the `mapCustomAttributes` method instead of the convert method?
> 
> The current implementation also has a bug when a pattern is not provided because we are assigning the updatedIndexAttributeValue to `""` and returning `""` if a pattern does not exist.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    protected static final String LOGSTASH_OPENSEARCH_INDEX_ATTRIBUTE_NAME = "index";
+    protected static final String REGEX_EXTRACTING_DATETIME_PATTERN = "%\\{(.*?)\\}";
+
+    @Override
+    protected void mapCustomAttributes(List<LogstashAttribute> logstashAttributes, LogstashAttributesMappings logstashAttributesMappings, Map<String, Object> pluginSettings) {
+
+        final LogstashAttribute logstashIndexAttribute = logstashAttributes.stream()
+                .filter(a -> a.getAttributeName().equals(LOGSTASH_OPENSEARCH_INDEX_ATTRIBUTE_NAME))
+                .findFirst()
+                .orElseThrow(() -> new LogstashConfigurationException("Index attribute not found"));
```
[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java:37]: I see, looks like I did not update this in the new revision, will do now. Thanks @cmanning09 @dlvenable

File: data-prepper-logstash-configuration/src/main/resources/org/opensearch/dataprepper/logstash/mapping/opensearch.mapping.yaml [hunk 9]
```diff
@@ -1,4 +1,5 @@
 pluginName: opensearch
+attributesMapperClass: org.opensearch.dataprepper.logstash.mapping.OpenSearchPluginAttributesMapper
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/main/resources/org/opensearch/dataprepper/logstash/mapping/opensearch.mapping.yaml:2]: This same code needs to be applied in `elasticsearch.mapping.yaml` and `amazon_es.mapping.yaml` as well.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        return new OpenSearchPluginAttributesMapper();
+    }
+
+    @Test
+    void convert_missing_indexAttribute_to_return_empty_pluginSettings() {
+
+        final LogstashAttribute logstashAttribute = mock(LogstashAttribute.class);
+        final LogstashAttributeValue logstashAttributeValue = mock(LogstashAttributeValue.class);
+
+        when(logstashAttribute.getAttributeName()).thenReturn("");
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:45]: This test is testing what happens when `index` is `null`. This is a good thing to test for sure.

But, these four lines are confusing. At first glance it looks like you are testing what happens when `index` is empty (`""`). These four lines are actually unnecessary and misleading.

We should also have a test for when the `index` is an empty string (`""`). You can test this using `createLogstashIndexAttribute("")` and don't need to copy these four lines around for it.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java (hunk 10 above)
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java]: Understood, added two testcases separating the two scenarios. Thanks for pointing it out.

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+        final LogstashAttribute logstashAttribute = createLogstashIndexAttribute("");
+
+        final LogstashAttributesMappings logstashAttributesMappings = mock(LogstashAttributesMappings.class);
+        when(logstashAttributesMappings.getMappedAttributeNames()).thenReturn(Collections.emptyMap());
+
+        final Map<String, Object> pluginSettings = createObjectUnderTest()
+                .mapAttributes(Collections.singletonList(logstashAttribute), logstashAttributesMappings);
+
+        assertThat(pluginSettings.size(), equalTo(1));
```
Comment (dlvenable) [inline data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:63]: I don't understand these assertions. There is one plugin setting here. But, it is not the `index`. What is it?

Comment (sshivanii) [reply data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:63]: `pluginSettings` is creating this: `pluginSettings: {null=}` where AttributeName is still `index` since that is a constant but AttributeValue is `""` so it's empty.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: @sshivanii, the current approach is not handling the default values for the [elasticsearch output plugin](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index) or the opensearch output plugin. We need to support a migration of these indices.

Regarding the opensearch output plugin: we need to dive deep here to understand what the default values our. The documentation is not clear at this time.

[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: @cmanning09 Regarding the default values for index in [Logstash ES output plugin](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index), it's dependent on the `ecs_compatibility` attribute which we do not support in our sink plugin so it will not be added to the converted `.yaml` file. Since we don't support ecs compatibility, I believe these default values should not be added.

I dug deeper into the OpenSearch code to find they only have a default index depending on `ecs` being enabled/disabled. Otherwise they throw an exception `Unsupported ECS compatibility`
https://github.com/opensearch-project/logstash-output-opensearch/blob/main/lib/logstash/outputs/opensearch.rb#L417

File: data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java (hunk 10 above)
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/test/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapperTest.java:57]: Thanks. Please change these values from empty strings to random strings. It will make the intent clearer.

File: data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: I'd like to suggest that when the index name is not supplied, Data Prepper uses the index name for when ECS is disabled. ECS compatibility is a broader topic than this PR is covering, and I don't think it should hinder this PR.

Thus, if the `index` is not present, the converter can just set the index to `logstash-%{uuuu.MM.dd}`. The current two tests which have no index value would both need to change to verify this new behavior.

[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: Also, please verify that this is also the default of the `amazon_es` plugin.

[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: I think the best solution here is to include a `defaultAttributes` property to the Logstash configuration mapping YAML. Other plugins may need to have default values where the Logstash default is different from the Data Prepper default.

As this problem exists in the current converter and this change introduces no regressions, I'd rather we wrap up this PR and support this more generically.

[the remark this answers is not in the document]
Comment (sshivanii) [reply data-prepper-logstash-configuration/src/main/java/org/opensearch/dataprepper/logstash/mapping/OpenSearchPluginAttributesMapper.java]: Checked the default index attribute value in `amazon_es` plugin and it is `logstash-%{+YYYY.MM.dd}` so the converted value will be `logstash-%{yyyy.MM.dd}`. This is different from the `elasticsearch` output plugin default, and the plugin name is not known at this time.
Reference: `https://github.com/awslabs/logstash-output-amazon_es/blob/master/lib/logstash/outputs/amazon_es/common_configs.rb#L17`

**Review by dlvenable — APPROVED** [review]
I believe this PR handles the current tasking and is ready to be merged. Please create a new GitHub issue for handling the default values for OpenSearch, Elasticsearch, and amazon_es index names.

**Review by cmanning09 — APPROVED** [review]
@sshivanii, please create an issue to track the delayed work for default values.

**Comment by sshivanii** [comment]
> @sshivanii, please create an issue to track the delayed work for default values.

Here's the issue for adding default values. https://github.com/opensearch-project/data-prepper/issues/1080
